### ISSUE TITLE
Remote Containers gets you throttled with Marketplace 

### ISSUE THREAD

**Issue by EricHripko** [microsoft/vscode-remote-release#2024]
- VSCode Version: 1.40.2
- Local OS Version: MacOS 10.14.5
- Remote OS Version: RHEL 7.7
- Remote Extension/Connection Type: Docker

Steps to Reproduce:

1. Work with a lot of projects using Remote Containers extension
2. Remote Containers downloads all extensions you specify
3. VSC Marketplace throttles you:
   ```
   Installing extensions...
   Installing extension 'ms-python.python' v2019.11.50794...
   Corrupt ZIP: end of central directory record signature not found
   ```

Does this issue occur when you try this locally?:
Yes, I can reproduce it by reaching out to the URL that (I believe) VSC uses to download the extensions.
```
$ curl https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/python/2019.11.50794/vspackage
{"$id":"1","innerException":null,"message":"Request was blocked due to exceeding usage of resource 'Count' in namespace 'AnonymousId'. For more information on why your request was blocked, see the topic \"Rate limits\" on the Microsoft Web site (https://go.microsoft.com/fwlink/?LinkId=823950).","typeName":"Microsoft.TeamFoundation.Framework.Server.RequestBlockedException, Microsoft.TeamFoundation.Framework.Server","typeKey":"RequestBlockedException","errorCode":0,"eventId":3000}
```

Does this issue occur when you try this locally and all extensions are disabled?:
No, since VSC doesn't download the extensions then.

VSC silently fails when an extension couldn't be downloaded - so I only find out it's broken when my IntelliSense or pylint is broken.
- Are there any solutions for this?
- Could Remote Containers pull extensions from the local installation?
- Could Remote Containers be configured to pull extensions from a custom URL (to vsix archive)?

**Comment by EricHripko** [comment]
Somewhat related to https://github.com/microsoft/vscode-remote-release/issues/185, as that can be one of solutions. Although I think it would be a more scalable solution if we could specify the URL to `vsix` in the `.devcontainer.json`.

**Comment by EricHripko** [comment]
Also somewhat related to https://github.com/microsoft/vscode-remote-release/issues/15

**Comment by EricHripko** [comment]
So, I've managed to get this to work by copying the VSIX to a local filesystem:
```dockerfile
# ...
COPY .devcontainer/ms-python.python.vsix /vscode/ms-python.python.vsix
```
and then changing `.devcontainer.json` to utilise a local file path:
```jsonc
/// ...
    "extensions": ["/vscode/ms-python.python.vsix"],
    "settings": {
/// ...
```

Is this a supported way to use the `.devcontainer.json`? This does enable me to avoid throttling and greatly speeds up the container rebuild (since it doesn't reach out for the network any more).

**Comment by chrmarti** [comment]
@sandy081 Is there anything Remote-Containers needs to do to avoid being throttled by the marketplace?

**Comment by EricHripko** [comment]
Thanks for picking this up @chrmarti! While we're waiting for @sandy081 to chip in, could you confirm if it's a supported use case to specify the path to `vsix` in `.devcontainer.json`?

**Comment by chrmarti** [comment]
@EricHripko My part of the code passes that string on to the server CLI unchanged. @sandy081 would know if paths to vsixes are supported.

**Comment by sandy081** [comment]
@chrmarti I do not think this is a throttling issue in remote. Because the remote logs are showing a different error. 

`Corrupt ZIP: end of central directory record signature not found`

It means the downloaded zip is not complete. It could be a network issue or an issue on Marketplace side.

remote cli do support installing from vsix, but I do not handle any path mappings. You have to make sure that the path you passed is valid and exists.

**Comment by EricHripko** [comment]
Hey folks! We've been using "Installing from .vsix" trick for some time now - and noticed that there appears to be an edge case with it in newer versions of VSC/RC extension:
```
[11099 ms] Installing extensions...
Extension 'cpptools-linux.vsix' was successfully installed.
Extension 'cmake-tools.vsix' was successfully installed.
[11099 ms] 
[11100 ms] Start: Run in container: # Copy /Users/<redacted>/.gnupg/trustdb.gpg to /root/.gnupg/trustdb.gpg
[11109 ms] 
[11109 ms] 
[11109 ms] Start: Run in container: ls /root/.vscode-server/extensionsCache
[11109 ms] Start: Run: gpg-connect-agent updatestartuptty /bye
[11115 ms] 
[11116 ms] ls: cannot access /root/.vscode-server/extensionsCache: No such file or directory
```

It looks like there's an assumption that `/root/.vscode-server/extensionsCache` exists. However, this doesn't appear to happen when extensions are installed from `.vsix`. The workarounds we found were:
- Opening Remote Containers again after it failed makes it work
- Changing `devcontainer.json` to instead install from Marketplace will correctly create `extensionsCache` folder. However, this quickly leads to throttling (Marketplace responds with rate limit responses).

Would you like me to submit a separate issue for that?

**Comment by mertaytore** [comment]
Hi! I'm running into the same issue @EricHripko just posted. I was able to track it down and identify that it works as expected until v0.119.0. and it breaks for 0.120.0+.

**Comment by chrmarti** [comment]
@EricHripko @mertaytore That should be fixed in 0.124.0. You get that with VS Code Insiders currently. (Fixed with #3269.)

**Comment by EricHripko** [comment]
That's awesome - thanks for the update, @chrmarti  🙌

**Comment by Rhahkeem** [comment]
Nice!! Was just coming here looking for this. Thanks!

**Comment by chrmarti** [comment]
@sandy081 Rereading the original description it seems the installer should detect the throttling case and not continue until "Corrupt ZIP".

Do we know why the throttling occurs? Maybe when a public IP address is shared between many machines that IP address can exceed the rate limit? Or @EricHripko are you just building a lot of images?

**Comment by EricHripko** [comment]
I use a corporate proxy for all external requests - your guess seems spot on 👍

**Comment by chrmarti** [comment]
Moving to vscode repo for not detecting the throttling case.
